test/junit/scala/collection/mutable/ArrayBuilderTest.scala

```diff
@@ -1,40 +1,35 @@
 package scala.collection.mutable
 
-import org.junit.Test
 import org.junit.Assert.assertEquals
+import org.junit.Test
+
+import scala.runtime.PStatics.VM_MaxArraySize
 import scala.tools.testkit.AssertUtil.assertThrows
 
-import scala.runtime.PStatics
 class ArrayBuilderTest {
-
-  /* Test for scala/bug#12617 */
-
-  val MaxArraySize: Int = PStatics.VM_MaxArraySize
-
   @Test
-  def testCapacityGrowthLimit: Unit = {
-
+  def t12617: Unit = {
     val ab: ArrayBuilder[Unit] = ArrayBuilder.make[Unit]
 
-    var i: Int = 0
-    while (i < MaxArraySize) {
-      ab.addOne(())
-      i += 1
-    }
+    // ArrayBuilder.ofUnit.addAll doesn't iterate if the iterator has a `knownSize`
+    ab.addAll(new Iterator[Unit] {
+      override def knownSize: Int = VM_MaxArraySize
+      def hasNext: Boolean = true
+      def next(): Unit = ()
+    })
```

Comment:
For those still watching, here's an interesting bit.

The while loop variant was running fine, actually very quickly because the JVM optimized the loop away. But as soon as I attached a debugger with a breakpoint somewhere below `addOne`, the VM had to de-optimize and the test didn't finish (or I guess it would have, eventually).

Reply:
The old joke is that the patient says doctor, it hurts when I do this. The doctor says then don't do that.

Collections still has a test problem, but arguably array ops would not work if array builder broke. Fortunately, array builder is simple (albeit the source is cut/paste for the special subclasses). For the purpose of building an array, it is preferable to array buffer, which is much more complex. Usually, you know how big the array must be, and growing it is an edge case. The convenience is in initializing the array.

Reply:
Tests are always welcome, there are certainly under-tested collections in the repo.

But yeah, we cannot just have tests that allocate Int.MaxValue elements, so testing for edge cases there is not easy.

> Is this a concern, or can we just not set breakpoints below `addOne`?

Note that I rewrote the test to no longer use a while loop.